### PR TITLE
nanocoap: bugfix token overflow

### PR DESCRIPTION
**Pull request by mjurczak**
### Contribution description

Fix for nanocoap read ouf of the input buffer:
https://github.com/RIOT-OS/RIOT/issues/14074

Corrected the options parsing loop condition to prevent skip-over the buffer end condition.
Added a pointer boundary check after adding token length declared in the packet header, but before making any access to the memory pointed by the current pkt_pos pointer.

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/issues/14074

### CODE REVIEW

**Comment by benpicco** [comment]
Thank you for the fix! Please prefix your commit messages with `nanocoap: …` to keep the history clean.
You can directly squash the code formatting commit.

**Comment by mjurczak** [comment]
No problem - fixed the commits.

File: sys/net/application_layer/nanocoap/nanocoap.c [hunk 1]
```diff
@@ -87,12 +87,17 @@ int coap_parse(coap_pkt_t *pkt, uint8_t *buf, size_t len)
         pkt->token = NULL;
     }
 
+    if (pkt_pos > pkt_end) {
```
Comment (kb2ma) [inline sys/net/application_layer/nanocoap/nanocoap.c:90]: Let's combine checks and save a few bytes. With your change to the while loop condition, we can move this check below the while loop, and remove the check currently at the end of the while loop. I'd change the debug message to something like "bad pkt len" to cover both cases. 

It also would be good to add a comment above, where testing the token length, that the validation is below the while loop.

**Comment by kb2ma** [comment]
I also retitled the PR. It makes it easier to find and collate.

File: sys/net/application_layer/nanocoap/nanocoap.c [hunk 2]
```diff
@@ -79,20 +79,25 @@ int coap_parse(coap_pkt_t *pkt, uint8_t *buf, size_t len)
...
+        pkt->token = NULL;
+    }
+    else if (coap_get_token_len(pkt) <= COAP_TOKEN_LENGTH_MAX) {
         pkt->token = pkt_pos;
+        /* pkt_pos range is validated after options parsing loop below */
         pkt_pos += coap_get_token_len(pkt);
     }
     else {
-        pkt->token = NULL;
+        DEBUG("tkl invalid\n");
```
Comment (miri64) [inline sys/net/application_layer/nanocoap/nanocoap.c:91]: Please prefix those DEBUG messages. Also, I would prefer to write out those abbreviations.

```suggestion
        DEBUG("nanocoap: token length invalid\n");
```

File: sys/net/application_layer/nanocoap/nanocoap.c [hunk 3]
```diff
@@ -129,14 +134,14 @@ int coap_parse(coap_pkt_t *pkt, uint8_t *buf, size_t len)
...
-
-            if (pkt_pos > (buf + len)) {
-                DEBUG("nanocoap: bad pkt\n");
-                return -EBADMSG;
-            }
         }
     }
 
+    if (pkt_pos > pkt_end) {
+        DEBUG("bad pkt len\n");
```
Comment (miri64) [inline sys/net/application_layer/nanocoap/nanocoap.c:141]:
```suggestion
        DEBUG("nanocoap: bad packet length\n");
```

**Comment by kb2ma** [comment]
@miri64, the PR has been updated to reflect your request on the debug messages. The squashed code is acceptable to me. However, the commit messages are unconventional and verbose. Please review them, and either follow-up or merge this PR.

**Comment by kb2ma** [comment]
@miri64 ping; see message above.

**Comment by mjurczak** [comment]
@kb2ma @miri64  Would you like me to modify the commit messages and make those more concise?

**Pull request merged**